File: fixtures/runtime/adapters/mcp/stdio-server.py
#!/usr/bin/env python3
import json
import os
import sys


BUFFER = b""
MAX_RESPONSE_BYTES = 1024 * 1024


def main() -> None:
    global BUFFER
    while True:
        chunk = os.read(sys.stdin.fileno(), 4096)
        if not chunk:
            return
        BUFFER += chunk
        parse_available_messages()


def parse_available_messages() -> None:
    global BUFFER
    while True:
        header_end = BUFFER.find(b"\r\n\r\n")
        if header_end == -1:
            return
        header = BUFFER[:header_end].decode("utf-8")
        content_length = None
        for line in header.splitlines():
            name, _, value = line.partition(":")
            if name.lower() == "content-length":
                content_length = int(value.strip())
                break
        if content_length is None:
            return
        body_start = header_end + 4
        body_end = body_start + content_length
        if len(BUFFER) < body_end:
            return
        body = BUFFER[body_start:body_end].decode("utf-8")
        BUFFER = BUFFER[body_end:]
        handle(json.loads(body))


def handle(request: dict[str, object]) -> None:
    request_id = request.get("id")
    if request_id is None:
        return
    method = request.get("method")
    if method == "initialize":
        respond(
            request_id,
            {
                "protocolVersion": "2025-06-18",
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "runx-rust-mcp-fixture", "version": "0.0.0"},
            },
        )
        return
    if method == "tools/list":
        respond(request_id, {"tools": tool_list()})
        return
    if method == "tools/call":
        handle_tool_call(request_id, request.get("params"))
        return
    respond_error(request_id, -32601, "method not found")


def tool_list() -> list[dict[str, object]]:
    return [
        tool("echo", "Echo a message through the fixture MCP server.", {"message": "string"}, ["message"]),
        tool("fail", "Return a fixture MCP error for testing.", {"message": "string"}, []),
        tool("sleep", "Never respond, for timeout testing.", {}, []),
        tool("env", "Return a single fixture server environment variable.", {"name": "string"}, ["name"]),
        tool("max-response", "Return a response body exactly at the MCP client size limit.", {}, []),
        tool("oversized-response", "Declare a response body over the MCP client size limit.", {}, []),
    ]


def tool(name: str, description: str, properties: dict[str, str], required: list[str]) -> dict[str, object]:
    return {
        "name": name,
        "description": description,
        "inputSchema": {
            "type": "object",
            "properties": {
                key: {"type": value, "description": f"{key}."} for key, value in properties.items()
            },
            "required": required,
            "additionalProperties": False,
        },
    }


def handle_tool_call(request_id: object, params: object) -> None:
    if not isinstance(params, dict) or not isinstance(params.get("name"), str):
        respond_error(request_id, -32602, "invalid tool call")
        return
    name = params["name"]
    args = params.get("arguments")
    if not isinstance(args, dict):
        args = {}
    if name == "sleep":
        return
    if name == "env":
        respond_text(request_id, os.environ.get(str(args.get("name", "")), ""))
        return
    if name == "fail":
        respond_error(request_id, -32000, f"fixture failure: {args.get('message', '')}")
        return
    if name == "max-response":
        respond_text(request_id, "x" * MAX_RESPONSE_BYTES)
        return
    if name == "oversized-response":
        respond_text(request_id, "x" * (MAX_RESPONSE_BYTES + 1))
        return
    if name != "echo":
        respond_error(request_id, -32601, "tool not found")
        return
    respond_text(request_id, str(args.get("message", "")))


def respond_text(request_id: object, text: str) -> None:
    respond(request_id, {"content": [{"type": "text", "text": text}]})


def respond(request_id: object, result: object) -> None:
    write({"jsonrpc": "2.0", "id": request_id, "result": result})


def respond_error(request_id: object, code: int, message: str) -> None:
    write({"jsonrpc": "2.0", "id": request_id, "error": {"code": code, "message": message}})


def write(message: dict[str, object]) -> None:
    body = json.dumps(message, separators=(",", ":")).encode("utf-8")
    sys.stdout.buffer.write(f"Content-Length: {len(body)}\r\n\r\n".encode("ascii") + body)
    sys.stdout.buffer.flush()


if __name__ == "__main__":
    main()

File: fixtures/runtime/adapters/mcp/stdio-server.mjs
import { appendFileSync } from "node:fs";

let input = Buffer.alloc(0);
const MAX_RESPONSE_BYTES = 1024 * 1024;
const startMarkerPath = process.env.RUNX_MCP_START_MARKER;
if (typeof startMarkerPath === "string" && startMarkerPath.length > 0) {
  appendLifecycle(startMarkerPath, "start");
}

process.stdin.on("data", (chunk) => {
  input = Buffer.concat([input, chunk]);
  parseAvailableMessages();
});

function parseAvailableMessages() {
  while (true) {
    const headerEnd = input.indexOf("\r\n\r\n");
    if (headerEnd === -1) {
      return;
    }

    const header = input.subarray(0, headerEnd).toString("utf8");
    const match = /Content-Length:\s*(\d+)/i.exec(header);
    if (!match) {
      return;
    }

    const bodyStart = headerEnd + 4;
    const bodyEnd = bodyStart + Number(match[1]);
    if (input.length < bodyEnd) {
      return;
    }

    const body = input.subarray(bodyStart, bodyEnd).toString("utf8");
    input = input.subarray(bodyEnd);
    handle(JSON.parse(body));
  }
}

function handle(request) {
  if (request.id === undefined) {
    return;
  }

  if (request.method === "initialize") {
    respond(request.id, {
      protocolVersion: "2025-06-18",
      capabilities: {
        tools: {},
      },
      serverInfo: {
        name: "runx-rust-mcp-fixture",
        version: "0.0.0",
      },
    });
    return;
  }

  if (request.method === "tools/list") {
    respond(request.id, {
      tools: [
        {
          name: "echo",
          description: "Echo a message through the fixture MCP server.",
          inputSchema: {
            type: "object",
            properties: {
              message: {
                type: "string",
                description: "Message to echo.",
              },
            },
            required: ["message"],
            additionalProperties: false,
          },
        },
        {
          name: "fail",
          description: "Return a fixture MCP error for testing.",
          inputSchema: {
            type: "object",
            properties: {
              message: {
                type: "string",
              },
            },
            additionalProperties: false,
          },
        },
        {
          name: "sleep",
          description: "Never respond, for timeout testing.",
          inputSchema: {
            type: "object",
            properties: {},
            additionalProperties: false,
          },
        },
        {
          name: "env",
          description: "Return a single fixture server environment variable.",
          inputSchema: {
            type: "object",
            properties: {
              name: {
                type: "string",
              },
            },
            required: ["name"],
            additionalProperties: false,
          },
        },
        {
          name: "max-response",
          description: "Return a response body exactly at the MCP client size limit.",
          inputSchema: {
            type: "object",
            properties: {},
            additionalProperties: false,
          },
        },
        {
          name: "oversized-response",
          description: "Declare a response body over the MCP client size limit.",
          inputSchema: {
            type: "object",
            properties: {},
            additionalProperties: false,
          },
        },
      ],
    });
    return;
  }

  if (request.method === "tools/call") {
    handleToolCall(request.id, request.params);
    return;
  }

  respondError(request.id, -32601, "method not found");
}

function handleToolCall(id, params) {
  if (!isRecord(params) || typeof params.name !== "string") {
    respondError(id, -32602, "invalid tool call");
    return;
  }

  const args = isRecord(params.arguments) ? params.arguments : {};

  if (params.name === "sleep") {
    startLifecycleHeartbeat(args.markerPath);
    return;
  }

  if (params.name === "env") {
    respond(id, {
      content: [
        {
          type: "text",
          text: String(process.env[String(args.name ?? "")] ?? ""),
        },
      ],
    });
    return;
  }

  if (params.name === "fail") {
    respondError(id, -32000, `fixture failure: ${String(args.message ?? "")}`);
    return;
  }

  if (params.name === "max-response") {
    respondWithTextBodyLength(id, MAX_RESPONSE_BYTES);
    return;
  }

  if (params.name === "oversized-response") {
    writeRaw(MAX_RESPONSE_BYTES + 1, "{}");
    return;
  }

  if (params.name !== "echo") {
    respondError(id, -32601, "tool not found");
    return;
  }

  respond(id, {
    content: [
      {
        type: "text",
        text: String(args.message ?? ""),
      },
    ],
  });
}

function respond(id, result) {
  write({
    jsonrpc: "2.0",
    id,
    result,
  });
}

function respondError(id, code, message) {
  write({
    jsonrpc: "2.0",
    id,
    error: {
      code,
      message,
    },
  });
}

function write(message) {
  const body = JSON.stringify(message);
  writeRaw(Buffer.byteLength(body, "utf8"), body);
}

function respondWithTextBodyLength(id, targetLength) {
  const empty = responseWithText(id, "");
  const emptyLength = Buffer.byteLength(JSON.stringify(empty), "utf8");
  const textLength = targetLength - emptyLength;
  if (textLength < 0) {
    throw new Error("target response length is too small");
  }
  const message = responseWithText(id, "x".repeat(textLength));
  const body = JSON.stringify(message);
  if (Buffer.byteLength(body, "utf8") !== targetLength) {
    throw new Error("sized fixture response length mismatch");
  }
  writeRaw(targetLength, body);
}

function responseWithText(id, text) {
  return {
    jsonrpc: "2.0",
    id,
    result: {
      content: [
        {
          type: "text",
          text,
        },
      ],
    },
  };
}

function writeRaw(contentLength, body) {
  process.stdout.write(`Content-Length: ${contentLength}\r\n\r\n${body}`);
}

function startLifecycleHeartbeat(markerPath) {
  if (typeof markerPath !== "string" || markerPath.length === 0) {
    return;
  }
  appendLifecycle(markerPath, "sleep-start");
  setInterval(() => appendLifecycle(markerPath, "heartbeat"), 25);
}

function appendLifecycle(markerPath, event) {
  appendFileSync(markerPath, `${event} ${process.pid} ${Date.now()}\n`);
}

function isRecord(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
